Add AlarmDecoder device info

tichris0 opened the pull request:
This commit makes AlarmDecoder operate as a proper device entity following the new model introduced a few years ago.

Code also has an internal dependency on a newer version of adext (>= 0.4.3) which has been updated correspondingly.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
AlarmDecoder will create new entity names through unique_ids. & proper component assignment  Users will most likely have to update any references to previously created AlarmDecoder entities.
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
The AlarmDecoder component has been rotting for a while now and never transitioned to the newer entity model.  This now properly defines it as a device integration and properly assigns all entities to it.  It is essentially a bug fix for the following long standing complaint: [https://github.com/home-assistant/core/issues/79599](https://github.com/home-assistant/core/issues/79599)


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: #79599
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

tichris0 committed:
Update AlarmDecoder component to newer model

This commit makes AlarmDecoder operate as a proper device entity following the new model introduced a few years ago.

Code also has an internal dependency on a newer version of adext (>= 0.4.3) which has been updated correspondingly.

tichris0 committed:
Merge branch 'dev' into dev

epenet commented on homeassistant/components/alarmdecoder/manifest.json:
```diff
@@ -3,8 +3,9 @@
   "iot_class": "local_push",
   "loggers": ["adext", "alarmdecoder"],
-  "requirements": ["adext==0.4.2"]
+  "requirements": ["adext==0.4.3"]
```
It is usually recommended that dependency bumps are done as standalone PR.
Only the changes directly linked to the dependency bump (for example when a property has been renamed) should go in the dependency bump PR.

tichris0 replied:
Thanks @epenet for taking the time to review my change.

Yes I saw that in the instructions, but wasn't sure this one made sense.  Curious to hear your feedback about this one and whether it's the right way to go.

Moreover, adext now exposes a new API to async/wait when it's done initializing.  The AlarmDecoder extension now wait for the object to be created before initializing.  The only way I can think to slice it would be to do the API bump and add the API async/wait call in AlarmDecoder and do no other change.  In isolation, that change would seem rather odd because there's currently no reason to wait for the initialization to happen given the old implementation.  It only really makes sense in light that I need now information from the initialization in order to create the entity in the extension.

If you believe that's still the way to go, I'd be happy to make the change accordingly, but it didn't quite obvious that it was the intention of the rule vs a simpler case of API bump where the API changes with no actual net effect on the code for example.

Please advise.

epenet replied:
It's still good practise, you can indicate that the bump is "just to keep to up-to-date", or that it is "preliminary for follow-up PRs that will make use of the functionnality".

tichris0 replied:
Ok no problem at all.  Here's the relevant PR that I submitted to handle the bump only:

[https://github.com/home-assistant/core/pull/117496](https://github.com/home-assistant/core/pull/117496)

epenet commented on homeassistant/components/alarmdecoder/sensor.py:
```diff
@@ -33,6 +43,15 @@ async def async_added_to_hass(self) -> None:
             )
         )
 
+    @property
+    def device_info(self):
+        return {
+            "identifiers": {(DOMAIN, self._client.serial_number)},
+            "manufacturer": "NuTech",
+            "serial_number": self._client.serial_number,
+            "sw_version": self._client.version_number,
+        }
+
```
That is a lot of duplicate code.
I suggest that you create a base entity in a new `entity.py` module.

```python
# in entity.py
class AlarmDecoderEntity(Entity):
    @property
    def device_info(self):
...
# in sensor.py
class AlarmDecoderSensor(AlarmDecoderEntity, SensorEntity):

```

tichris0 committed:
Created AlarmDecoder entity

Added an alarmdecoder entity so the device_info can be re-used across the integration

tichris0 committed:
Merge branch 'dev' of https://github.com/tichris0/HAcore into dev

epenet commented on homeassistant/components/alarmdecoder/manifest.json:
```diff
@@ -3,8 +3,9 @@
   "name": "AlarmDecoder",
   "codeowners": [],
   "config_flow": true,
+  "integration_type": "device",
```
Make this a standalone PR: see #117519

epenet commented on homeassistant/components/alarmdecoder/alarm_control_panel.py:
```diff
@@ -75,10 +76,11 @@ async def async_setup_entry(
     """Representation of an AlarmDecoder-based alarm panel."""
 
     _attr_name = "Alarm Panel"
+    _attr_has_entity_name = True
```
Should be moved to the base entity
```suggestion
```

epenet commented on homeassistant/components/alarmdecoder/binary_sensor.py:
```diff
@@ -53,20 +57,22 @@ async def async_setup_entry(
+class AlarmDecoderBinarySensor(AlarmDecoderEntity, BinarySensorEntity):
     """Representation of an AlarmDecoder binary sensor."""
 
+    _attr_has_entity_name = True
```
Should be moved to the base entity
```suggestion
```

epenet commented on homeassistant/components/alarmdecoder/binary_sensor.py:
```diff
@@ -76,6 +82,8 @@ def __init__(
         relay_chan,
     ):
         """Initialize the binary_sensor."""
+        self._attr_unique_id = f"{client.serial_number}-zone-{zone_number}"
+        self._client = client
```
```suggestion
        super().__init__(client)
        self._attr_unique_id = f"{client.serial_number}-zone-{zone_number}"
```

epenet commented on homeassistant/components/alarmdecoder/alarm_control_panel.py:
```diff
@@ -89,11 +91,13 @@ class AlarmDecoderAlarmPanel(AlarmControlPanelEntity):
 
     def __init__(self, client, auto_bypass, code_arm_required, alt_night_mode):
         """Initialize the alarm panel."""
+        self._attr_unique_id = f"{client.serial_number}-panel"
         self._client = client
```
```suggestion
        super().__init__(client)
        self._attr_unique_id = f"{client.serial_number}-panel"
```

epenet commented on homeassistant/components/alarmdecoder/entity.py:
```diff
@@ -0,0 +1,18 @@
+    DOMAIN,
+)
+
+class AlarmDecoderEntity(Entity):
+    """Define a base AlarmDecoder entity."""
+
+    @property
+    def device_info(self):
+        return {
+            "identifiers": {(DOMAIN, self._client.serial_number)},
+            "manufacturer": "NuTech",
+            "serial_number": self._client.serial_number,
+            "sw_version": self._client.version_number,
+        }
```
```suggestion
class AlarmDecoderEntity(Entity):
    """Define a base AlarmDecoder entity."""
    
    ```suggestion
class AlarmDecoderEntity(Entity):
    """Define a base AlarmDecoder entity."""

    _attr_has_entity_name = True
    
    def __init__(self, client):
        self._client = client
        self._attr_device_info = DeviceInfo(
            identifiers={(DOMAIN, self._client.serial_number)},
            manufacturer="NuTech",
            serial_number=self._client.serial_number,
            sw_version=self._client.version_number,
        )
```

epenet replied:
I think you can also set `entry_type` on the device_info

epenet commented on homeassistant/components/alarmdecoder/sensor.py:
```diff
@@ -6,25 +6,36 @@
     _attr_should_poll = False
 
+    def __init__(self, client):
+        self._attr_unique_id = f"{client.serial_number}-display"
+        self._client = client
```
```suggestion
        super().__init__(client)
        self._attr_unique_id = f"{client.serial_number}-display"
```

epenet reviewed (changes requested):
Please rebase this PR now that the dependency bump has gone through

tichris0 committed:
Merge branch 'home-assistant:dev' into dev

tichris0 committed:
Move _attr_has_entity_name to base entity

As per code review suggestion, clean up the object model.

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

tichris0 committed:
Missed one suggestion with the prior commit

Moves _attr_has_entity_name to base entity

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

tichris0 committed:
Address some ruff issues

tichris0 committed:
Apply additional ruff cleanups

Ran ruff again to clean up a few files tat weren't picked up last time

epenet commented on homeassistant/components/alarmdecoder/sensor.py:
```diff
@@ -6,25 +6,33 @@
+class AlarmDecoderSensor(AlarmDecoderEntity, SensorEntity):
     """Representation of an AlarmDecoder keypad."""
 
+    _attr_has_entity_name = True
```
Not needed I think, as it comes from the base entity.
```suggestion
```

tichris0 replied:
You're correct.  I verified that it still works with this change

epenet commented on homeassistant/components/alarmdecoder/entity.py:
```diff
@@ -0,0 +1,24 @@
+
+    def __init__(self, client):
+        """Initialize the alarm decoder entity."""
+        self._client = client
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, self._client.serial_number)},
+            manufacturer="NuTech",
+            serial_number=self._client.serial_number,
+            sw_version=self._client.version_number,
+        )
```
I think actually we don't need to make `_client` accessible anymore, as it was only necessary for the `device_info`.
```suggestion
        self._attr_device_info = DeviceInfo(
            identifiers={(DOMAIN, client.serial_number)},
            manufacturer="NuTech",
            serial_number=client.serial_number,
            sw_version=client.version_number,
        )
```

tichris0 replied:
It's used for more than just the registration.  I need adext functionality across child entities.  As a reference, see alarm_control_panel which calls a couple functions on it.

epenet replied:
Indeed!
We should still use the local variable directly inside attr_device_info:
```suggestion
        self._client = client
        self._attr_device_info = DeviceInfo(
            identifiers={(DOMAIN, client.serial_number)},
            manufacturer="NuTech",
            serial_number=client.serial_number,
            sw_version=client.version_number,
        )
```

tichris0 replied:
Yup that seems fair & more readable.

epenet commented on homeassistant/components/alarmdecoder/entity.py:
```diff
@@ -0,0 +1,24 @@
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity import Entity
+
+from .const import (
+    DOMAIN,
+)
```
```suggestion
from .const import DOMAIN
```

tichris0 replied:
Weird ruff didn't catch this one.

tichris0 committed:
Apply suggestions from code review

Some additional cleanup of style & removal of unnecessary code

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

epenet reviewed (changes requested):
I adjusted my code suggestion regarding `_attr_device_info`
Please adjust accordingly

epenet commented on homeassistant/components/alarmdecoder/manifest.json:
```diff
@@ -3,6 +3,7 @@
   "name": "AlarmDecoder",
   "codeowners": [],
   "config_flow": true,
+  "integration_type": "device",
```
By changing the integration type here, it is breaking hassfest.
Please ensure that you run `python3 -m script.hassfest`

tichris0 replied:
I did run this on my side, but I didn't see any files changed as a consequence of this.  Do you have any insight into which file I'm supposed to pick up?

tichris0 replied:
Any insight into what is tripping it up?

epenet replied:
Ah! indeed.
The issues seems to be that you didn't insert it in the right place.
It should be ordered and moved under "documentation"
Then script should run correctly.

epenet replied:
You actually have to run it twice in your case:
- once to sort the manifest (which will fix the incorrect order)
- then a second time to update the `homeassistant/generated/integrations.json` generated file

tichris0 replied:
Ah yes perfect.  That's it.  It was generated the second time and I see it now :)  Thanks.

tichris0 committed:
Properly generated the integration file

generation had to happen twice for this to work.  Now that it's generated, I'm including the missing update.

tichris0 committed:
Apply suggestions from code review

Use local client variable instead of self._client

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

tichris0 committed:
Sort the manifest

documentation was added, but it wasn't sorted properly in the key/value pairs

tichris0 committed:
Merge branch 'dev' of https://github.com/tichris0/HAcore into dev

epenet reviewed (changes requested):
Almost there...
Please add entity file to coverage ignore file (coveragerc)

tichris0 committed:
Add alarmdecoder entity file to coverage ignore file

Added the alarmdecoder entity file so it is ignored for coverage

tichris0 commented:
> Almost there... Please add entity file to coverage ignore file (coveragerc)

There we go all green :).  Thanks for holding my hand through this.  Should be a lot smoother next time.

epenet reviewed (approved):
LGTM 👍

epenet commented:
Please mark as "Ready for review" when ready to merge.

tichris0 committed:
Merge branch 'dev' into dev

epenet commented:
Thanks  @tichris0 
You should consider a couple of follow ups:
- add yourself as code owner 
- move away from hass.data and use entry.runtime_data

joostlek commented:
Adding unique id isn't considered a breaking change as far as I'm aware

tichris0 commented:
> Adding unique id isn't considered a breaking change as far as I'm aware

Yah I think you may be right.  When I tried the old model and switched to the new just now, the entities seemed to have still worked, so it doesn't look like a breaking change after all :)

joostlek commented:
Another nice follow-up is btw to add entity translations :)

tichris0 commented:
> Thanks @tichris0 You should consider a couple of follow ups:
> 
>     * add yourself as code owner
> 
>     * move away from hass.data and use entry.runtime_data

> Another nice follow-up is btw to add entity translations :)

Thanks I'll look into those.
